File: shared-kmp/src/commonMain/kotlin/ai/aligned/net/dto/Dtos.kt
package ai.aligned.net.dto

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

// ─── Health / settings ──────────────────────────────────────────────────────

@Serializable data class HealthDto(
    val status: String, val version: String, val storyCount: Int,
    val lastUpdated: String, val services: Services
) { @Serializable data class Services(val twitter: Boolean, val openai: Boolean, val tavily: Boolean, val cerebras: Boolean) }

@Serializable data class SettingsResponseDto(
    val settings: SettingsDto,
    val availableModels: List<ModelOption> = emptyList(),
    val availableLanguages: List<LanguageOption> = emptyList()
)
@Serializable data class SettingsDto(
    val synthesisModel: String, val chatModel: String,
    val language: String, val customXApiKey: String, val algorithm: String
)
@Serializable data class ModelOption(val id: String, val label: String)
@Serializable data class LanguageOption(val id: String, val label: String, val dir: String)

// ─── Categories ─────────────────────────────────────────────────────────────

@Serializable data class CategoriesDto(
    val categories: List<CategoryDto> = emptyList(),
    val custom: List<CategoryDto>     = emptyList(),
    val all: List<CategoryDto>        = emptyList()
)
@Serializable data class CategoryDto(
    val id: String, val label: String, val color: String,
    val bgColor: String? = null, val section: String? = null
)

// ─── News / stories / tweets ────────────────────────────────────────────────

@Serializable data class NewsDto(val stories: List<StoryDto> = emptyList())

@Serializable data class StoryDto(
    val id: String, val headline: String, val summary: String,
    val category: String,
    val tweets: List<TweetDto> = emptyList(),
    val citations: List<JsonElement> = emptyList(),
    val createdAt: String, val updatedAt: String,
    val tweetCount: Int = 0, val totalEngagement: Long = 0,
    val signals: Signals? = null
) { @Serializable data class Signals(
        val topScore: Int = 0,
        val categories: List<String> = emptyList(),
        val flags: JsonObject = JsonObject(emptyMap())
   )
}

@Serializable data class TweetDto(
    val id: String, val text: String,
    val authorName: String, val authorUsername: String,
    val authorProfileImage: String, val authorFollowers: Long = 0,
    val authorLocation: String? = null,
    val createdAt: String,
    val likes: Long = 0, val retweets: Long = 0, val replies: Long = 0,
    val bookmarks: Long = 0, val views: Long = 0, val quotes: Long = 0,
    val url: String,
    val media: List<TweetMedia> = emptyList()
)
@Serializable data class TweetMedia(
    val type: String, val url: String,
    val width: Int = 0, val height: Int = 0
)

// ─── Search ─────────────────────────────────────────────────────────────────

@Serializable data class SearchDto(
    val stories: List<StoryDto> = emptyList(),
    val query: String = ""
)

// ─── Lists (editorial roll-up) ──────────────────────────────────────────────

@Serializable data class ListsDto(
    val overview: ListsOverview? = null,
    val groups: List<GroupDto> = emptyList(),
    val historyDates: List<String> = emptyList()
)
@Serializable data class ListsOverview(
    val timestamp: String, val date: String, val execSummary: String,
    val execBriefing: ExecBriefing? = null,
    val groupSummaries: List<GroupSummary> = emptyList(),
    val topStories: List<TopStoryRef> = emptyList()
)
@Serializable data class ExecBriefing(
    val leadHeadline: String, val leadGroupId: String, val leadBody: String,
    val sections: List<BriefingSection> = emptyList()
)
@Serializable data class BriefingSection(val subhead: String, val groupId: String, val body: String)
@Serializable data class GroupSummary(
    val groupId: String, val groupName: String, val color: String,
    val summary: String, val storyCount: Int
)
@Serializable data class TopStoryRef(
    val headline: String, val groupId: String,
    val sourceHandle: String, val engagement: Long, val tweetUrl: String
)
@Serializable data class GroupDto(
    val groupId: String, val groupName: String, val timestamp: String,
    val execSummary: String, val storyCount: Int,
    val subcategories: List<SubcategoryDto> = emptyList(),
    val topStories: List<GroupStoryDto>     = emptyList()
)
@Serializable data class SubcategoryDto(
    val id: String, val name: String, val summary: String,
    val storyCount: Int, val stories: List<GroupStoryDto> = emptyList()
)
@Serializable data class GroupStoryDto(
    val id: String, val headline: String, val summary: String,
    val analysis: String, val sourceHandle: String,
    val tweets: List<TweetDto> = emptyList(),
    val engagement: Long = 0, val createdAt: String
)

// ─── Map / Events / Accounts / History ──────────────────────────────────────

@Serializable data class MapDto(val markers: List<MapMarker> = emptyList())
@Serializable data class MapMarker(
    val lat: Double, val lng: Double, val city: String, val country: String,
    val groupId: String, val groupName: String, val groupColor: String,
    val stories: List<MapStoryRef> = emptyList()
)
@Serializable data class MapStoryRef(val headline: String, val sourceHandle: String, val tweetUrl: String)

@Serializable data class EventsDto(val events: List<EventDto> = emptyList())
@Serializable data class EventDto(
    val id: String, val name: String, val date: String, val location: String,
    val category: String, val description: String, val source: EventSource? = null
)
@Serializable data class EventSource(val headline: String, val authorUsername: String, val tweetUrl: String)

@Serializable data class AccountsDto(val accounts: List<AccountDto> = emptyList())
@Serializable data class AccountDto(
    val id: String, val username: String, val name: String,
    val profileImage: String, val followers: Long,
    val description: String, val addedAt: String
)

@Serializable data class HistoryEntryDto(
    val id: String, val type: String, val query: String,
    val confidence: Int = 0, val insightCount: Int = 0, val completedAt: String
)

// ─── Feedback ───────────────────────────────────────────────────────────────

@Serializable data class FeedbackRequest(val storyId: String, val category: String, val vote: String)
@Serializable data class FeedbackPostDto(val feedback: FeedbackEcho)
@Serializable data class FeedbackEcho(
    val storyId: String, val headline: String, val category: String,
    val sentiment: String, val timestamp: String
)
@Serializable data class FeedbackStatsDto(
    val total: Int, val thumbsUp: Int, val thumbsDown: Int,
    val byCategory: Map<String, FeedbackCount> = emptyMap()
)
@Serializable data class FeedbackCount(val up: Int = 0, val down: Int = 0)

// ─── Summarize ──────────────────────────────────────────────────────────────

@Serializable data class TweetInput(val text: String, val authorUsername: String)
@Serializable data class SummarizeRequest(val tweets: List<TweetInput>)
@Serializable data class SummarizeDto(val overview: String, val individual: List<String> = emptyList())

// ─── Chat (SSE) ─────────────────────────────────────────────────────────────

@Serializable data class ChatRequest(
    val message: String,
    val storyId: String? = null,
    val history: List<ChatTurn> = emptyList()
)
@Serializable data class ChatTurn(val role: String, val content: String)

@Serializable sealed class ChatEvent {
    @Serializable @SerialName("token")
    data class Token(val text: String) : ChatEvent()
    @Serializable @SerialName("done")
    object Done : ChatEvent()
}

// ─── Research ───────────────────────────────────────────────────────────────

@Serializable data class ResearchStartRequest(val query: String)
@Serializable data class ResearchStartDto(val id: String, val status: String)

@Serializable data class ResearchDto(
    val id: String, val query: String, val status: String,
    val steps: List<ResearchStep> = emptyList(),
    val currentStep: Int = 0,
    val insights: List<Insight> = emptyList(),
    val sourceNetwork: List<JsonElement> = emptyList(),
    val activityLog: List<String> = emptyList(),
    val tweets: List<TweetDto> = emptyList(),
    val startedAt: String? = null,
    val completedAt: String? = null,
    val summaryAnswer: String? = null
)
@Serializable data class ResearchStep(val name: String, val status: String, val detail: String? = null)
@Serializable data class Insight(
    val title: String, val summary: String,
    val confidence: Int = 0, val sources: List<String> = emptyList()
)

File: windows-app/Aligned.Core/Net/Dto/Dtos.cs
using System.Text.Json.Serialization;

namespace Aligned.Core.Net.Dto;

// XAML binding (x:DataType) requires regular get/set properties, not init-only.
// All DTOs are mutable classes with default values so they round-trip JSON ⇄ XAML.

// ─── Health / settings ──────────────────────────────────────────────────────

public class HealthDto
{
    public string Status { get; set; } = "";
    public string Version { get; set; } = "";
    public int StoryCount { get; set; }
    public string LastUpdated { get; set; } = "";
    public ServicesDto Services { get; set; } = new();
    public class ServicesDto
    {
        public bool Twitter { get; set; }
        public bool OpenAi { get; set; }
        public bool Tavily { get; set; }
        public bool Cerebras { get; set; }
    }
}

public class SettingsResponseDto
{
    public SettingsDto Settings { get; set; } = new();
    public List<ModelOption> AvailableModels { get; set; } = new();
    public List<LanguageOption> AvailableLanguages { get; set; } = new();
}
public class SettingsDto
{
    public string SynthesisModel { get; set; } = "";
    public string ChatModel { get; set; } = "";
    public string Language { get; set; } = "en";
    public string CustomXApiKey { get; set; } = "";
    public string Algorithm { get; set; } = "";
}
public class ModelOption    { public string Id { get; set; } = ""; public string Label { get; set; } = ""; }
public class LanguageOption { public string Id { get; set; } = ""; public string Label { get; set; } = ""; public string Dir { get; set; } = "ltr"; }

// ─── Categories ─────────────────────────────────────────────────────────────

public class CategoriesDto
{
    public List<CategoryDto> Categories { get; set; } = new();
    public List<CategoryDto> Custom { get; set; } = new();
    public List<CategoryDto> All { get; set; } = new();
}
public class CategoryDto
{
    public string Id { get; set; } = "";
    public string Label { get; set; } = "";
    public string Color { get; set; } = "#999999";
    public string? BgColor { get; set; }
    public string? Section { get; set; }
}

// ─── News / story / tweet ───────────────────────────────────────────────────

public class NewsDto { public List<StoryDto> Stories { get; set; } = new(); }

public class StoryDto
{
    public string Id { get; set; } = "";
    public string Headline { get; set; } = "";
    public string Summary { get; set; } = "";
    public string Category { get; set; } = "";
    public List<TweetDto> Tweets { get; set; } = new();
    public string CreatedAt { get; set; } = "";
    public string UpdatedAt { get; set; } = "";
    public int TweetCount { get; set; }
    public long TotalEngagement { get; set; }
}

public class TweetDto
{
    public string Id { get; set; } = "";
    public string Text { get; set; } = "";
    [JsonPropertyName("authorName")]         public string AuthorName { get; set; } = "";
    [JsonPropertyName("authorUsername")]     public string AuthorUsername { get; set; } = "";
    [JsonPropertyName("authorProfileImage")] public string AuthorProfileImage { get; set; } = "";
    [JsonPropertyName("authorFollowers")]    public long AuthorFollowers { get; set; }
    public string CreatedAt { get; set; } = "";
    public long Likes { get; set; }
    public long Retweets { get; set; }
    public long Replies { get; set; }
    public long Bookmarks { get; set; }
    public long Views { get; set; }
    public long Quotes { get; set; }
    public string Url { get; set; } = "";
    public List<TweetMedia> Media { get; set; } = new();
}

public class TweetMedia
{
    public string Type { get; set; } = "";
    public string Url { get; set; } = "";
    public int Width { get; set; }
    public int Height { get; set; }
}

// ─── Search ─────────────────────────────────────────────────────────────────

public class SearchDto
{
    public List<StoryDto> Stories { get; set; } = new();
    public string Query { get; set; } = "";
}

// ─── Lists ──────────────────────────────────────────────────────────────────

public class ListsDto
{
    public ListsOverview? Overview { get; set; }
    public List<GroupDto> Groups { get; set; } = new();
    public List<string> HistoryDates { get; set; } = new();
}

public class ListsOverview
{
    public string Timestamp { get; set; } = "";
    public string Date { get; set; } = "";
    public string ExecSummary { get; set; } = "";
    public ExecBriefing? ExecBriefing { get; set; }
    public List<GroupSummary> GroupSummaries { get; set; } = new();
    public List<TopStoryRef> TopStories { get; set; } = new();
}
public class ExecBriefing
{
    public string LeadHeadline { get; set; } = "";
    public string LeadGroupId { get; set; } = "";
    public string LeadBody { get; set; } = "";
    public List<BriefingSection> Sections { get; set; } = new();
}
public class BriefingSection { public string Subhead { get; set; } = ""; public string GroupId { get; set; } = ""; public string Body { get; set; } = ""; }
public class GroupSummary
{
    public string GroupId { get; set; } = "";
    public string GroupName { get; set; } = "";
    public string Color { get; set; } = "";
    public string Summary { get; set; } = "";
    public int StoryCount { get; set; }
}
public class TopStoryRef
{
    public string Headline { get; set; } = "";
    public string GroupId { get; set; } = "";
    public string SourceHandle { get; set; } = "";
    public long Engagement { get; set; }
    public string TweetUrl { get; set; } = "";
}
public class GroupDto
{
    public string GroupId { get; set; } = "";
    public string GroupName { get; set; } = "";
    public string Timestamp { get; set; } = "";
    public string ExecSummary { get; set; } = "";
    public int StoryCount { get; set; }
    public List<SubcategoryDto> Subcategories { get; set; } = new();
    public List<GroupStoryDto> TopStories { get; set; } = new();
}
public class SubcategoryDto
{
    public string Id { get; set; } = "";
    public string Name { get; set; } = "";
    public string Summary { get; set; } = "";
    public int StoryCount { get; set; }
    public List<GroupStoryDto> Stories { get; set; } = new();
}
public class GroupStoryDto
{
    public string Id { get; set; } = "";
    public string Headline { get; set; } = "";
    public string Summary { get; set; } = "";
    public string Analysis { get; set; } = "";
    public string SourceHandle { get; set; } = "";
    public List<TweetDto> Tweets { get; set; } = new();
    public long Engagement { get; set; }
    public string CreatedAt { get; set; } = "";
}

// ─── Map / Events / Accounts / History ──────────────────────────────────────

public class MapDto { public List<MapMarker> Markers { get; set; } = new(); }
public class MapMarker
{
    public double Lat { get; set; }
    public double Lng { get; set; }
    public string City { get; set; } = "";
    public string Country { get; set; } = "";
    public string GroupId { get; set; } = "";
    public string GroupName { get; set; } = "";
    public string GroupColor { get; set; } = "";
    public List<MapStoryRef> Stories { get; set; } = new();
}
public class MapStoryRef { public string Headline { get; set; } = ""; public string SourceHandle { get; set; } = ""; public string TweetUrl { get; set; } = ""; }

public class EventsDto { public List<EventDto> Events { get; set; } = new(); }
public class EventDto
{
    public string Id { get; set; } = "";
    public string Name { get; set; } = "";
    public string Date { get; set; } = "";
    public string Location { get; set; } = "";
    public string Category { get; set; } = "";
    public string Description { get; set; } = "";
    public EventSource? Source { get; set; }
}
public class EventSource { public string Headline { get; set; } = ""; public string AuthorUsername { get; set; } = ""; public string TweetUrl { get; set; } = ""; }

public class AccountsDto { public List<AccountDto> Accounts { get; set; } = new(); }
public class AccountDto
{
    public string Id { get; set; } = "";
    public string Username { get; set; } = "";
    public string Name { get; set; } = "";
    public string ProfileImage { get; set; } = "";
    public long Followers { get; set; }
    public string Description { get; set; } = "";
    public string AddedAt { get; set; } = "";
}

public class HistoryEntryDto
{
    public string Id { get; set; } = "";
    public string Type { get; set; } = "";
    public string Query { get; set; } = "";
    public int Confidence { get; set; }
    public int InsightCount { get; set; }
    public string CompletedAt { get; set; } = "";
}

// ─── Feedback ───────────────────────────────────────────────────────────────

public class FeedbackRequest
{
    public string StoryId { get; set; } = "";
    public string Category { get; set; } = "";
    public string Vote { get; set; } = "up";
    public FeedbackRequest() { }
    public FeedbackRequest(string storyId, string category, string vote)
    { StoryId = storyId; Category = category; Vote = vote; }
}
public class FeedbackPostDto { public FeedbackEcho Feedback { get; set; } = new(); }
public class FeedbackEcho
{
    public string StoryId { get; set; } = "";
    public string Headline { get; set; } = "";
    public string Category { get; set; } = "";
    public string Sentiment { get; set; } = "";
    public string Timestamp { get; set; } = "";
}
public class FeedbackStatsDto
{
    public int Total { get; set; }
    public int ThumbsUp { get; set; }
    public int ThumbsDown { get; set; }
    public Dictionary<string, FeedbackCount> ByCategory { get; set; } = new();
}
public class FeedbackCount
{
    public int Up { get; set; }
    public int Down { get; set; }
}

// ─── Summarize ──────────────────────────────────────────────────────────────

public class TweetInput
{
    public string Text { get; set; } = "";
    public string AuthorUsername { get; set; } = "";
    public TweetInput() { }
    public TweetInput(string text, string authorUsername) { Text = text; AuthorUsername = authorUsername; }
}
public class SummarizeRequest
{
    public List<TweetInput> Tweets { get; set; } = new();
    public SummarizeRequest() { }
    public SummarizeRequest(List<TweetInput> tweets) { Tweets = tweets; }
}
public class SummarizeDto { public string Overview { get; set; } = ""; public List<string> Individual { get; set; } = new(); }

// ─── Chat (SSE) ─────────────────────────────────────────────────────────────

public class ChatRequest
{
    public string Message { get; set; } = "";
    public string? StoryId { get; set; }
    public List<ChatTurn> History { get; set; } = new();
    public ChatRequest() { }
    public ChatRequest(string message, string? storyId, List<ChatTurn>? history)
    { Message = message; StoryId = storyId; History = history ?? new(); }
}
public class ChatTurn
{
    public string Role { get; set; } = "user";
    public string Content { get; set; } = "";
    public ChatTurn() { }
    public ChatTurn(string role, string content) { Role = role; Content = content; }
}

public abstract record ChatEvent
{
    public sealed record Token(string Text) : ChatEvent;
    public sealed record Done() : ChatEvent;
}

// ─── Research ───────────────────────────────────────────────────────────────

public class ResearchStartRequest
{
    public string Query { get; set; } = "";
    public ResearchStartRequest() { }
    public ResearchStartRequest(string query) { Query = query; }
}
public class ResearchStartDto
{
    public string Id { get; set; } = "";
    public string Status { get; set; } = "";
}

public class ResearchDto
{
    public string Id { get; set; } = "";
    public string Query { get; set; } = "";
    public string Status { get; set; } = "";
    public List<ResearchStep> Steps { get; set; } = new();
    public int CurrentStep { get; set; }
    public List<Insight> Insights { get; set; } = new();
    public List<string> ActivityLog { get; set; } = new();
    public List<TweetDto> Tweets { get; set; } = new();
    public string? StartedAt { get; set; }
    public string? CompletedAt { get; set; }
    public string? SummaryAnswer { get; set; }
}
public class ResearchStep
{
    public string Name { get; set; } = "";
    public string Status { get; set; } = "";
    public string? Detail { get; set; }
}
public class Insight
{
    public string Title { get; set; } = "";
    public string Summary { get; set; } = "";
    public int Confidence { get; set; }
    public List<string> Sources { get; set; } = new();
}

public enum Vote { Up, Down }
public static class VoteExtensions { public static string Wire(this Vote v) => v == Vote.Up ? "up" : "down"; }
